packages/errors/src/message-formatter.ts

```diff
@@ -0,0 +1,21 @@
+import { SolanaErrorCode } from './codes';
+import { SolanaErrorMessages } from './messages';
+
+export function getHumanReadableErrorMessage<TErrorCode extends SolanaErrorCode>(
+    code: TErrorCode,
+    context: object = {},
+): string {
+    const messageFormatString = SolanaErrorMessages[code];
+    const message = messageFormatString.replace(/(?<!\\)\$(\w+)/g, (substring, variableName) =>
+        variableName in context ? `${context[variableName as keyof typeof context]}` : substring,
+    );
+    return message;
```

Comment:
I think this is great and will be enough 95% of the time.

However, how would you feel about, additionally, **allowing `SolanaErrorMessages[code]` to be a function** of type `(context: object) => string`. That way, if `typeof SolanaErrorMessages[code] === "string"`, we keep the same logic, otherwise, we simply call the custom function to convert a context from a specific error into a message.

This feature can be added later on without breaking changes so probably best to not future-proof too early but I do think this could be useful in some situations.

Reply:
What you've described is 1000% the way I wrote it before this revision, and it was typesafe, and it was beautiful.

Then I realized that if they're functions:

* The function implementation could, itself, fatal, depending on what inputs you send it, which is really really bad when you're trying to throw an error.
* This makes it essentially impossible to serialize the inputs for use with the error decoder two PRs later in this stack. If you have, say, a `CryptoKey` in your `context` and your error message does something useful and informative with it, then you go to production and your error gets minified, there's nothing I can do now to serialize that for decoding by `npx @solana/errors decode ...`

So I rolled that beautiful idea back and made them Just Format Strings™.

Reply:
Ah yeah that's fair enough. I'm happy to leave things as-is and if we ever encounter a scenario where templated strings is not enough, we can revisit at this point. 👌